protofsm/state_machine_test.go

```diff
@@ -0,0 +1,330 @@
+package protofsm
+
+import (
+	"encoding/hex"
+	"fmt"
+	"sync/atomic"
+	"testing"
+
+	"github.com/btcsuite/btcd/btcec/v2"
+	"github.com/btcsuite/btcd/wire"
+	"github.com/lightningnetwork/lnd/fn"
+	"github.com/lightningnetwork/lnd/lnwire"
+	"github.com/stretchr/testify/mock"
+	"github.com/stretchr/testify/require"
+)
+
+type dummyEvents interface {
+	dummy()
+}
+
+type goToFin struct {
+}
+
+func (g *goToFin) dummy() {
+}
+
+type emitInternal struct {
+}
+
+func (e *emitInternal) dummy() {
+}
+
+type daemonEvents struct {
+}
+
+func (s *daemonEvents) dummy() {
+}
+
+type dummyEnv struct {
+	mock.Mock
+}
+
+func (d *dummyEnv) CleanUp() error {
+	args := d.Called()
+
+	return args.Error(0)
+}
+
+type dummyStateStart struct {
+	canSend *atomic.Bool
+}
+
+var (
+	hexDecode = func(keyStr string) []byte {
+		keyBytes, _ := hex.DecodeString(keyStr)
+		return keyBytes
+	}
+	pub1, _ = btcec.ParsePubKey(hexDecode(
+		"02ec95e4e8ad994861b95fc5986eedaac24739e5ea3d0634db4c8ccd44cd" +
+			"a126ea",
+	))
+	pub2, _ = btcec.ParsePubKey(hexDecode(
+		"0356167ba3e54ac542e86e906d4186aba9ca0b9df45001c62b753d33fe06" +
+			"f5b4e8",
+	))
+)
+
+func (d *dummyStateStart) ProcessEvent(event dummyEvents, env *dummyEnv,
```

Comment:
This could also use `mock` so we don't end up in a situation where we'd ask the question - do you write tests to test your test code😂

Reply:
So I started with that, but then found it easier to write the state machine here, as a sort of documentation, and also IMO easier to handle the autonomous execution, like when I test that it can loop back on itself and handle internal emitted events, etc.